pallets/parachain-staking/src/migrations.rs

```diff
@@ -32,24 +35,87 @@ pub enum StakingStorageVersion {
 	V6,     // Fix delegator replacement bug
 }
 
-#[cfg(feature = "try-runtime")]
-impl StakingStorageVersion {
-	/// The latest storage version.
-	#[allow(dead_code)]
-	fn latest() -> Self {
-		Self::V6
-	}
-}
+// Migration for ParachainStaking pallet to mutate CandidatePool to be
+// a CountedStorageMap and switch to new Pallet StorageVersion paradigm.
+pub mod v7 {
+	use super::*;
+	use crate::{CandidatePool, Config, Pallet};
+
+	#[cfg(feature = "try-runtime")]
+	use frame_support::traits::GetStorageVersion;
+	use frame_support::{
+		generate_storage_alias,
+		traits::{Get, OnRuntimeUpgrade, StorageVersion as NewStorageVersion},
+		weights::Weight,
+	};
+	use log::info;
+	#[cfg(feature = "try-runtime")]
+	use sp_runtime::traits::Zero;
+	use sp_std::marker::PhantomData;
+
+	// Get storage items into scope which are removed during this migration
+	generate_storage_alias!(ParachainStaking, CandidateCount => Value<u32>);
+	generate_storage_alias!(ParachainStaking, StorageVersion => Value<StakingStorageVersion>);
+
+	pub struct ParachainStakingMigrationV7<T: Config>(PhantomData<T>);
+
+	impl<T: Config> OnRuntimeUpgrade for ParachainStakingMigrationV7<T> {
+		fn on_runtime_upgrade() -> Weight {
+			// migrate CandidateCount
+			frame_support::migration::remove_storage_prefix(b"ParachainStaking", b"CandidateCount", &[]);
+			let candidate_count = CandidatePool::<T>::initialize_counter();
+
+			// migrate StorageVersion to new paradigm
+			frame_support::migration::remove_storage_prefix(b"ParachainStaking", b"StorageVersion", &[]);
+			NewStorageVersion::new(7).put::<Pallet<T>>();
+
+			info!("💰 completed parachain staking migration to v7 ✅",);
+			T::DbWeight::get().reads_writes(candidate_count.into(), candidate_count.saturating_add(3).into())
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			assert!(
+				!CandidateCount::get().unwrap().is_zero(),
+				"CandidateCount already migrated"
+			);
+			assert!(CandidatePool::<T>::count().is_zero(), "Candidate counter already set");
+			assert!(StorageVersion::get() == Some(StakingStorageVersion::V6));
+
+			assert!(CandidatePool::<T>::initialize_counter() == CandidatePool::<T>::iter().count() as u32);
+
+			info!("💰 parachain staking migration to v7 passes PRE migrate checks ✅",);
+			Ok(())
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade() -> Result<(), &'static str> {
+			// check count
+			assert!(
+				!frame_support::migration::have_storage_value(b"StakePallet", b"CandidateCount", &[]),
```

Comment:
Shouldn't we check using the old pallet name?

Reply:
What do you mean by old pallet name? The pallet name has never been changed?

Reply:
Isnt't the first parameter the name of the pallet? Is that `ParachainStaking` or `StakePallet`?

Reply:
You are right, I had tunnel vision. `StakePallet` is only used in the unit test 🙈 Great catch! I will derive the pallet name directly from storage so that the migration will be correct independent of the corresponding runtime.

Reply:
1866d37a58f36b18fc57def17555a1e47ac3e60a